src/py/flwr/client/connection/rere_fleet_connection.py

```diff
@@ -0,0 +1,242 @@
+from flwr.proto.fab_pb2 import GetFabRequest, GetFabResponse  # pylint: disable=E0611
+from flwr.proto.fleet_pb2 import (  # pylint: disable=E0611
+    CreateNodeRequest,
+    CreateNodeResponse,
+    DeleteNodeRequest,
+    PingRequest,
+    PingResponse,
+    PullTaskInsRequest,
+    PushTaskResRequest,
+)
+from flwr.proto.node_pb2 import Node  # pylint: disable=E0611
+from flwr.proto.run_pb2 import GetRunRequest, GetRunResponse  # pylint: disable=E0611
+from flwr.proto.task_pb2 import TaskIns  # pylint: disable=E0611
+
+from .fleet_api import FleetApi
+from .fleet_connection import FleetConnection
+
+
+class RereFleetConnection(FleetConnection):
+    """Network-based request-response fleet connection."""
+
+    def __init__(  # pylint: disable=R0913, R0914, R0915
+        self,
+        server_address: str,
+        insecure: bool,
+        retry_invoker: RetryInvoker,
+        max_message_length: int = GRPC_MAX_MESSAGE_LENGTH,  # pylint: disable=W0613
+        root_certificates: bytes | str | None = None,
+        authentication_keys: (
+            tuple[ec.EllipticCurvePrivateKey, ec.EllipticCurvePublicKey] | None
+        ) = None,
+    ) -> None:
+        """Initialize the RereFleetConnection."""
+        super().__init__(
+            server_address=server_address,
+            insecure=insecure,
+            retry_invoker=retry_invoker,
+            max_message_length=max_message_length,
+            root_certificates=root_certificates,
+            authentication_keys=authentication_keys,
+        )
+        self.msg_id_to_metadata: dict[str, Metadata] = {}
+        self.node: Node | None = None
+        self.ping_thread: threading.Thread | None = None
+        self.ping_stop_event = threading.Event()
+        self.channel: grpc.Channel | None = None
+
+    @property
+    @abstractmethod
+    def api(self) -> FleetApi:
+        """The proxy providing low-level access to the Fleet API server."""
+
+    def ping(self) -> None:
+        """Ping the SuperLink."""
+        # Get Node
+        if self.node is None:
+            log(ERROR, "Node instance missing")
+            return
+
+        # Construct the ping request
+        req = PingRequest(node=self.node, ping_interval=PING_DEFAULT_INTERVAL)
+
+        # Call FleetAPI
+        res: PingResponse = self.api.Ping(req, timeout=PING_CALL_TIMEOUT)
+
+        # Check if success
+        if not res.success:
+            raise RuntimeError("Ping failed unexpectedly.")
+
+        # Wait
+        rd = random.uniform(*PING_RANDOM_RANGE)
+        next_interval: float = PING_DEFAULT_INTERVAL - PING_CALL_TIMEOUT
+        next_interval *= PING_BASE_MULTIPLIER + rd
+        if not self.ping_stop_event.is_set():
+            self.ping_stop_event.wait(next_interval)
+
+    def create_node(self) -> int | None:
+        """Request to create a node."""
+        # Call FleetAPI
+        req = CreateNodeRequest(ping_interval=PING_DEFAULT_INTERVAL)
+        res: CreateNodeResponse = self.retrier.invoke(
```

Comment:
I made a related minor comment in #4054 whether we should call this `self.retry_invoker`.